src/Microsoft.AspNet.Http.Abstractions/Extensions/UseMiddlewareExtensions.cs

```diff
@@ -24,22 +24,36 @@ public static IApplicationBuilder UseMiddleware(this IApplicationBuilder builder
             {
                 var instance = ActivatorUtilities.CreateInstance(builder.ApplicationServices, middleware, new[] { next }.Concat(args).ToArray());
                 var methodinfo = middleware.GetMethod("Invoke", BindingFlags.Instance | BindingFlags.Public);
+                if (methodinfo == null)
+                {
+                    throw new InvalidOperationException("No public Invoke method found.");
+                }
+
+                if (!typeof(Task).IsAssignableFrom(methodinfo.ReturnType))
+                {
+                    throw new InvalidOperationException("Invoke does not return an object of type Task.");
+                }
+
                 var parameters = methodinfo.GetParameters();
-                if (parameters[0].ParameterType != typeof(HttpContext))
+
+                if (parameters.Length == 0 || parameters[0].ParameterType != typeof(HttpContext))
                 {
-                    throw new Exception("Middleware Invoke method must take first argument of HttpContext");
+                    throw new InvalidOperationException("The Invoke method's first argument must be of type HttpContext.");
```

Comment:
I'm fine either way. But should definitely use either `nameof()` or `typeof(X).FullName`.